File: examples/nirfmx/wlangen/csharp/VS2010/80211GenerateWaveform(MultiUserOFDMA)/cs/Program.cs
using System;
using NationalInstruments.RFToolkits.Wlan.Generation;
using NationalInstruments.ModularInstruments.NIRfsg;
using NationalInstruments.ModularInstruments.NIRfsg.Internal;
using System.Runtime.InteropServices;


namespace _80211GenerateWaveformMultiUserOFDMA
{
   class Program
   {
      static void Main(string[] args)
      {
         GenerateWaveformMultipleRFSGExample example = new GenerateWaveformMultipleRFSGExample();
         example.Run();
      }
   }

   class GenerateWaveformMultipleRFSGExample
   {
      const int MAX_WLAN_CHANNELS = 4;
      int numTx, mappingMatrixType, PPDUType, numberOfUsers, guardIntervalType;

      double carrierFrequency;
      double channelBandwidth;
      double[] powerLevel = new double[MAX_WLAN_CHANNELS] { -10.0, -10.0, -10.0, -10.0 };
      double[] externalAttenuation = new double[MAX_WLAN_CHANNELS] { 0, 0, 0, 0 };

      int[] RuSize = {niWLANGConstants.RuSize26,niWLANGConstants.RuSize26,niWLANGConstants.RuSize52,
                        niWLANGConstants.RuSize26,niWLANGConstants.RuSize106};
      int[] RuOffsetMruIndex = { 0, 1, 2, 4, 5 };
      int[] McsIndex = { 0, 0, 0, 0, 0 };
      int[] numberOfSpaceTimeStream = { 1, 1, 1, 1, 1 };
      int[] payloadLength = { 100, 100, 100, 100, 100 };

      NIRfsg[] rfsgSessions = new NIRfsg[MAX_WLAN_CHANNELS];
      niWLANG wlanSession;
      NIRfsg externalLOSession = null;
      String[] rfsgResourceName = new string[MAX_WLAN_CHANNELS] { "RIO0", "RIO1", "RIO2", "RIO3" };
      String rfsgClockSource = RfsgFrequencyReferenceSource.PxiClock;

      String waveformName = "Wlan";
      String channelString;
      int[] triggerLines = { 0, 1 };

      public void Run()
      {
         try
         {
            initGlobalVaribales();
            ConfigureWlanGenerationSession();
            ConfigureRfsgSession();
            CreateAndDownloadWaveform();
            StopGeneration();
         }
         catch (Exception e)
         {
            Console.WriteLine("Error : " + e.ToString());
            Console.WriteLine("press any key to exit");
            Console.ReadKey();
         }
         finally
         {
            if (wlanSession != null)
               wlanSession.Close();
         }
      }

      private void initGlobalVaribales()
      {
         channelBandwidth = 20e+6;
         numTx = 1;
         mappingMatrixType = niWLANGConstants.MappingMatrixTypeDirect;
         PPDUType = niWLANGConstants.PpduTypeMuPpdu;
         numberOfUsers = 5;
         guardIntervalType = niWLANGConstants.GuardIntervalTypeOneByFour;

         carrierFrequency = 5.18e+9;
      }

      private void ConfigureWlanGenerationSession()
      {
         int isNewSession;

         if (wlanSession == null)
            wlanSession = new niWLANG("WLANG", niWLANGConstants.CompatibilityVersion060000, out isNewSession);

         wlanSession.SetStandard(null, niWLANGConstants.Standard80211AxMimoOfdm);
         wlanSession.SetChannelBandwidth(null, channelBandwidth);
         wlanSession.SetNumberOfTransmitChannels(null, numTx);
         wlanSession.SetMappingMatrixType(null, mappingMatrixType);
         wlanSession.SetPPDUType(null, PPDUType);
         wlanSession.SetNumberOfUsers(null, numberOfUsers);
         wlanSession.SetOFDMGuardIntervalType(null, guardIntervalType);

         for (int i = 0; i < numberOfUsers; i++)
         {
            channelString = "user" + i;
            wlanSession.SetRUSize(channelString, RuSize[i]);
            wlanSession.SetRUOffsetMruIndex(channelString, RuOffsetMruIndex[i]);
            wlanSession.SetMcsIndex(channelString, McsIndex[i]);
            wlanSession.SetNumberOfSpaceTimeStreams(channelString, numberOfSpaceTimeStream[i]);
            channelString = channelString + "/mpdu0";
            wlanSession.SetPayloadDataLength(channelString, payloadLength[i]);
         }

         wlanSession.SetRFBlankingEnabled(null, niWLANGConstants.True);
      }

      private void CreateAndDownloadWaveform()
      {
         double iqRate, waveformDuration;
         int iqWaveformSize;
         double[] actualHeadRoom = new double[numTx];
         String script = @"script GenerateWlan
				                repeat forever
					                generate Wlan
				                end repeat
			                end script";

         IntPtr[] rfsgHandle = new IntPtr[numTx];
         IntPtr externalLOHandle = new IntPtr();

         for (int i = 0; i < numTx; i++)
            rfsgHandle[i] = rfsgSessions[i].GetInstrumentHandle().DangerousGetHandle();
         if (externalLOSession != null)
            externalLOHandle = externalLOSession.GetInstrumentHandle().DangerousGetHandle();

         wlanSession.ConfigureMultipleDeviceSynchronization(rfsgHandle, numTx, rfsgClockSource, triggerLines, triggerLines.Length);
         wlanSession.RFSGConfigureFrequencySingleLO(rfsgHandle, niWLANGConstants.LOSourceOnboard, externalLOHandle, carrierFrequency, niWLANGConstants.False, niWLANGConstants.False);

         wlanSession.RFSGCreateAndDownloadMIMOWaveforms(rfsgHandle, null, numTx, waveformName);
         /*Configure Script*/
         for (int i = 0; i < numTx; i++)
         {
            niWLANG.WLANG_RFSGConfigureScript(rfsgHandle[i], null, script, powerLevel[i]);
         }
         wlanSession.RFSGMultipleDeviceInitiate(rfsgHandle);

         //Check for successful generation
         CheckGeneration();

         wlanSession.GetIqRate(String.Empty, out iqRate);
         wlanSession.GetIqWaveformSize(String.Empty, out iqWaveformSize);
         waveformDuration = iqWaveformSize / iqRate;

         Console.WriteLine("Waveform Duration : {s}" + waveformDuration);
         Console.WriteLine("Actual HeadRoom\n");
         for (int i = 0; i < numTx; i++)
         {
            channelString = "Channel" + i;
            wlanSession.GetActualHeadroom(channelString, out actualHeadRoom[i]);
            Console.WriteLine("\t" + actualHeadRoom[i]);
         }

         Console.WriteLine("Press any key to exit");
         Console.ReadKey();
      }

      private void ConfigureRfsgSession()
      {
         //Open all RFSG Sessions
         for (int i = 0; i < numTx; i++)
         {
            if (rfsgSessions[i] == null)
            {
               rfsgSessions[i] = new NIRfsg(rfsgResourceName[i], false, true);
            }
            rfsgSessions[i].RF.PowerLevelType = RfsgRFPowerLevelType.PeakPower;
            rfsgSessions[i].Arb.GenerationMode = RfsgWaveformGenerationMode.Script;
            rfsgSessions[i].RF.ExternalGain = -(externalAttenuation[i]);
         }
      }

      private bool CheckGeneration()
      {
         RfsgGenerationStatus status = RfsgGenerationStatus.InProgress;
         for (int i = 0; i < numTx; i++)
         {
            status = rfsgSessions[i].CheckGenerationStatus();
            if (status == RfsgGenerationStatus.Complete)
               break;
         }
         return Convert.ToBoolean(status);
      }

      private void StopGeneration()
      {
         for (int i = 0; i < numTx; i++)
         {
            if (rfsgSessions[i] != null)
            {
               rfsgSessions[i].Abort();
               rfsgSessions[i].RF.OutputEnabled = false;
               rfsgSessions[i].Utility.Commit();
               niWLANG.WLANG_RFSGClearDatabase(rfsgSessions[i].GetInstrumentHandle().DangerousGetHandle(), "", waveformName);
               rfsgSessions[i].Close();
            }
         }
      }
   }
}

File: examples/nirfmx/wlangen/csharp/VS2010/80211GenerateWaveform(MultiUserOFDMA)/vb/Program.vb
Imports NationalInstruments.RFToolkits.Wlan.Generation
Imports NationalInstruments.ModularInstruments.NIRfsg
Imports NationalInstruments.ModularInstruments.NIRfsg.Internal
Imports System.Runtime.InteropServices


Namespace _80211GenerateWaveformMultiUserOFDMA
   Class Program
      Friend Shared Sub Main(args As String())
         Dim example As New GenerateWaveformMultipleRFSGExample()
         example.Run()
      End Sub
   End Class

   Class GenerateWaveformMultipleRFSGExample

      Const MAX_WLAN_CHANNELS As Integer = 4
      Private numTx As Integer, mappingMatrixType As Integer, PPDUType As Integer, numberOfUsers As Integer, guardIntervalType As Integer

      Private carrierFrequency As Double
      Private channelBandwidth As Double
      Private powerLevel As Double() = New Double(MAX_WLAN_CHANNELS - 1) {-10.0, -10.0, -10.0, -10.0}
      Private externalAttenuation As Double() = New Double(MAX_WLAN_CHANNELS - 1) {0, 0, 0, 0}

      Private RuSize As Integer() = {niWLANGConstants.RuSize26, niWLANGConstants.RuSize26, niWLANGConstants.RuSize52, niWLANGConstants.RuSize26, niWLANGConstants.RuSize106}
      Private RuOffsetMruIndex As Integer() = {0, 1, 2, 4, 5}
      Private McsIndex As Integer() = {0, 0, 0, 0, 0}
      Private numberOfSpaceTimeStream As Integer() = {1, 1, 1, 1, 1}
      Private payloadLength As Integer() = {100, 100, 100, 100, 100}

      Private rfsgSessions As NIRfsg() = New NIRfsg(MAX_WLAN_CHANNELS - 1) {}
      Private wlanSession As niWLANG
      Private externalLOSession As NIRfsg = Nothing
      Private rfsgResourceName As [String]() = New String(MAX_WLAN_CHANNELS - 1) {"RIO0", "RIO1", "RIO2", "RIO3"}
      Private rfsgClockSource As [String] = RfsgFrequencyReferenceSource.PxiClock

      Private waveformName As [String] = "Wlan"
      Private channelString As [String]
      Private triggerLines As Integer() = {0, 1}

      Public Sub Run()
         Try
            initGlobalVaribales()
            ConfigureWlanGenerationSession()
            ConfigureRfsgSession()
            CreateAndDownloadWaveform()
            StopGeneration()
         Catch e As Exception
            Console.WriteLine("Error : " & e.ToString())
            Console.WriteLine("press any key to exit")
            Console.ReadKey()
         Finally
            If wlanSession IsNot Nothing Then
               wlanSession.Close()
            End If
         End Try
      End Sub

      Private Sub initGlobalVaribales()
         channelBandwidth = 20000000.0
         numTx = 1
         mappingMatrixType = niWLANGConstants.MappingMatrixTypeDirect
         PPDUType = niWLANGConstants.PpduTypeMuPpdu
         numberOfUsers = 5
         guardIntervalType = niWLANGConstants.GuardIntervalTypeOneByFour

         carrierFrequency = 5180000000.0
      End Sub

      Private Sub ConfigureWlanGenerationSession()
         Dim isNewSession As Integer

         If wlanSession Is Nothing Then
            wlanSession = New niWLANG("WLANG", niWLANGConstants.CompatibilityVersion060000, isNewSession)
         End If

         wlanSession.SetStandard(Nothing, niWLANGConstants.Standard80211AxMimoOfdm)
         wlanSession.SetChannelBandwidth(Nothing, channelBandwidth)
         wlanSession.SetNumberOfTransmitChannels(Nothing, numTx)
         wlanSession.SetMappingMatrixType(Nothing, mappingMatrixType)
         wlanSession.SetPPDUType(Nothing, PPDUType)
         wlanSession.SetNumberOfUsers(Nothing, numberOfUsers)
         wlanSession.SetOFDMGuardIntervalType(Nothing, guardIntervalType)

         For i As Integer = 0 To numberOfUsers - 1
            channelString = "user" & i
            wlanSession.SetRUSize(channelString, RuSize(i))
            wlanSession.SetRUOffsetMruIndex(channelString, RuOffsetMruIndex(i))
            wlanSession.SetScalarAttributeI32(channelString, niWLANGProperties.McsIndex, McsIndex(i))
            wlanSession.SetNumberOfSpaceTimeStreams(channelString, numberOfSpaceTimeStream(i))
            channelString = channelString & "/mpdu0"
            wlanSession.SetPayloadDataLength(channelString, payloadLength(i))
         Next

         wlanSession.SetRFBlankingEnabled(Nothing, niWLANGConstants.[True])
      End Sub

      Private Sub CreateAndDownloadWaveform()
         Dim iqRate As Double, waveformDuration As Double
         Dim iqWaveformSize As Integer
         Dim actualHeadRoom As Double() = New Double(numTx - 1) {}
         Dim script As [String] = "script GenerateWlan" & vbCr & vbLf & vbTab & vbTab & vbTab & vbTab & "                repeat forever" & vbCr & vbLf & vbTab & vbTab & vbTab & vbTab & vbTab & "                generate Wlan" & vbCr & vbLf & vbTab & vbTab & vbTab & vbTab & "                end repeat" & vbCr & vbLf & vbTab & vbTab & vbTab & "                end script"

         Dim rfsgHandle As IntPtr() = New IntPtr(numTx - 1) {}
         Dim externalLOHandle As New IntPtr()

         For i As Integer = 0 To numTx - 1
            rfsgHandle(i) = rfsgSessions(i).GetInstrumentHandle().DangerousGetHandle()
         Next
         If externalLOSession IsNot Nothing Then
            externalLOHandle = externalLOSession.GetInstrumentHandle().DangerousGetHandle()
         End If

         wlanSession.ConfigureMultipleDeviceSynchronization(rfsgHandle, numTx, rfsgClockSource, triggerLines, triggerLines.Length)
         wlanSession.RFSGConfigureFrequencySingleLO(rfsgHandle, niWLANGConstants.LOSourceOnboard, externalLOHandle, carrierFrequency, niWLANGConstants.[False], niWLANGConstants.[False])

         wlanSession.RFSGCreateAndDownloadMIMOWaveforms(rfsgHandle, Nothing, numTx, waveformName)
         'Configure Script

         For i As Integer = 0 To numTx - 1
            niWLANG.WLANG_RFSGConfigureScript(rfsgHandle(i), Nothing, script, powerLevel(i))
         Next
         wlanSession.RFSGMultipleDeviceInitiate(rfsgHandle)
         'Check for successful generation
         CheckGeneration()
         wlanSession.GetIqRate([String].Empty, iqRate)
         wlanSession.GetIqWaveformSize([String].Empty, iqWaveformSize)
         waveformDuration = iqWaveformSize / iqRate

         Console.WriteLine("Waveform Duration : {s}" & waveformDuration)
         Console.WriteLine("Actual HeadRoom" & vbLf)
         For i As Integer = 0 To numTx - 1
            channelString = "Channel" & i
            wlanSession.GetActualHeadroom(channelString, actualHeadRoom(i))
            Console.WriteLine(vbTab & actualHeadRoom(i))
         Next
         Console.WriteLine("Press any key to exit")
         Console.ReadKey()

      End Sub

      Private Sub ConfigureRfsgSession()
         'Open all RFSG Sessions
         For i As Integer = 0 To numTx - 1
            If rfsgSessions(i) Is Nothing Then
               rfsgSessions(i) = New NIRfsg(rfsgResourceName(i), False, True)
            End If
            rfsgSessions(i).RF.PowerLevelType = RfsgRFPowerLevelType.PeakPower
            rfsgSessions(i).Arb.GenerationMode = RfsgWaveformGenerationMode.Script
            rfsgSessions(i).RF.ExternalGain = -(externalAttenuation(i))
         Next
      End Sub

      Private Function CheckGeneration() As Boolean
         Dim status As RfsgGenerationStatus = RfsgGenerationStatus.InProgress
         For i As Integer = 0 To numTx - 1
            status = rfsgSessions(i).CheckGenerationStatus()
            If status = RfsgGenerationStatus.Complete Then
               Exit For
            End If
         Next
         Return Convert.ToBoolean(status)
      End Function

      Private Sub StopGeneration()
         For i As Integer = 0 To numTx - 1
            If rfsgSessions(i) IsNot Nothing Then
               rfsgSessions(i).Abort()
               rfsgSessions(i).RF.OutputEnabled = False
               rfsgSessions(i).Utility.Commit()
               niWLANG.WLANG_RFSGClearDatabase(rfsgSessions(i).GetInstrumentHandle().DangerousGetHandle(), "", waveformName)
               rfsgSessions(i).Close()
            End If
         Next
      End Sub
   End Class
End Namespace
